### ISSUE TITLE
Lazy Registration

### ISSUE THREAD

**Issue by seesharper** [seesharper/LightInject#21]
For very large applications, it might not be a good solution to register all services up front. This can cause significant delay during startup and we need to provide some sort of lazy registration where services are registered when requested.

We need to hook into the process of resolving unknown services.

When an unknown service is requested, we will scan the assembly where the requested service type is contained. It will as usual look for composition roots and execute them if any are found in the assembly. Otherwise we will do a simple scan and register services using the standard conventions.

We also need to look into the mocking extensions as we might need to disable overwriting existing service registrations. 

The next challenge is when the service type and the implementing type are contained in two different assemblies. The interface assembly will not have a reference to the implementing assembly. This could be solved by using an assembly attribute that is used to decorate the service assembly. This attribute gives LightInject a hint on the composition root type to be executed. This could be in the form of a Type or the type name. 

```
public class CompositionRootAttribute : Attribute
{        
    public CompositionRootAttribute(Type compositionRootType)
    {

    }

    public CompositionRootAttribute(string typeName)
    {

    }

}
```

**Comment by bounav** [comment]
Hello,

I'm trying to get lazy registration to work on a project with many Composition Roots spread across multiple assemblies...

The documentation says LightInject "will scanning the assembly where the service is contained" (see extract below).

> Lazy Composition Roots
>LightInject is capable of registering services on a need to have basis. For a large application that has a lot of services, it might not be the best solution to register all these services up front as this could seriously hurt the startup time of our application due to extensive assembly loading.
>
>If an unregistered service is requested, LightInject will scan the assembly where this service is contained.

_How does it know which assemblies to scan?_

I ended up adding `[assembly: CompositionRootType(typeof(RelevantCompositionRoot))]` into each assemblies's Composition Root file.

Then I called `Container.RegisterAssembly("Foo*.dll");`  to tell LightInject to load `FooOne.dll`, `FooTwo.dll`, etc.

Is this right approach?

If so maybe the documentation could be updated to clarify that use case a little?

Thanks,

Ben
